File: ios/Classes/document/OcrRunner.swift
import UIKit
import Vision

/// A recognized word plus its normalized `[0..1]` top-left bounding box — the
/// unit of the invisible text layer drawn by the searchable-PDF exporter.
/// v1.2 Phase DC8.
struct SupyOcrWord {
  let text: String
  let left: Double
  let top: Double
  let width: Double
  let height: Double
}

/// Runs `VNRecognizeTextRequest` over a set of pages and concatenates the
/// recognized strings.
///
/// Unlike the Android side, Vision text recognition supports `ar` in addition
/// to Latin scripts, so the iOS pipeline is the language-complete one for
/// Supy's coverage matrix.
final class OcrRunner {

  /// Background queue for VNRequest work. Vision is CPU-heavy; never run on
  /// the main thread.
  private let workQueue = DispatchQueue(
    label: "io.supy.scanner.document.ocr",
    qos: .userInitiated
  )

  /// Runs OCR over each `(uri, image)` pair on a background queue and
  /// invokes [completion] on the main queue once. Per-page failures yield
  /// empty text for that page rather than failing the whole batch.
  func run(
    pages: [(uri: URL, image: UIImage, width: Int, height: Int, quality: String?, qualityScore: Double?)],
    languages: [String],
    completion: @escaping (_ pages: [[String: Any]], _ ocrText: String) -> Void
  ) {
    guard !pages.isEmpty else {
      DispatchQueue.main.async { completion([], "") }
      return
    }

    let longEdgeCap = SupyDeviceTier.detect().ocrLongEdgeCap

    workQueue.async {
      let recognized: [String] = pages.map { page in
        let prepared = Self.downscaleIfNeeded(page.image, longEdgeCap: longEdgeCap)
        return Self.recognizeText(in: prepared, languages: languages)
      }
      let entries: [[String: Any]] = pages.map { page in
        var entry: [String: Any] = [
          "uri": page.uri.absoluteString,
          "width": page.width,
          "height": page.height,
        ]
        if let q = page.quality { entry["quality"] = q }
        if let qs = page.qualityScore { entry["qualityScore"] = qs }
        return entry
      }
      let joined = recognized
        .joined(separator: "\n\n")
        .trimmingCharacters(in: .whitespacesAndNewlines)
      DispatchQueue.main.async {
        completion(entries, joined)
      }
    }
  }

  /// Like [run], but also returns each page's word boxes so a single OCR pass
  /// feeds both `ocrText` and the searchable-PDF invisible text layer — no
  /// second recognition pass. `wordsPerPage` is index-aligned with [pages].
  /// Word boxes are normalized `[0..1]` (top-left origin), so the downscale
  /// applied for recognition doesn't affect them. v1.2 Phase DC8.
  func runWithWords(
    pages: [(uri: URL, image: UIImage, width: Int, height: Int, quality: String?, qualityScore: Double?)],
    languages: [String],
    completion: @escaping (_ pages: [[String: Any]], _ ocrText: String, _ wordsPerPage: [[SupyOcrWord]]) -> Void
  ) {
    guard !pages.isEmpty else {
      DispatchQueue.main.async { completion([], "", []) }
      return
    }

    let longEdgeCap = SupyDeviceTier.detect().ocrLongEdgeCap

    workQueue.async {
      var recognized: [String] = []
      var wordsPerPage: [[SupyOcrWord]] = []
      recognized.reserveCapacity(pages.count)
      wordsPerPage.reserveCapacity(pages.count)
      for page in pages {
        let prepared = Self.downscaleIfNeeded(page.image, longEdgeCap: longEdgeCap)
        let (text, words) = Self.recognizeTextAndWords(in: prepared, languages: languages)
        recognized.append(text)
        wordsPerPage.append(words)
      }
      let entries: [[String: Any]] = pages.map { page in
        var entry: [String: Any] = [
          "uri": page.uri.absoluteString,
          "width": page.width,
          "height": page.height,
        ]
        if let q = page.quality { entry["quality"] = q }
        if let qs = page.qualityScore { entry["qualityScore"] = qs }
        return entry
      }
      let joined = recognized
        .joined(separator: "\n\n")
        .trimmingCharacters(in: .whitespacesAndNewlines)
      DispatchQueue.main.async {
        completion(entries, joined, wordsPerPage)
      }
    }
  }

  /// Runs a single-image structured OCR pass and returns the block → line →
  /// element tree consumed by `SupyRecognizedText.fromMap` on the Dart side.
  ///
  /// Vision has no paragraph/block concept, so all recognized lines are
  /// wrapped in one synthetic block whose box is the union of its lines.
  /// Bounding boxes are normalized `[0..1]` with the origin flipped from
  /// Vision's bottom-left to Supy's top-left convention.
  func recognizeStructured(
    image: UIImage,
    languages: [String],
    includeElements: Bool,
    completion: @escaping (_ tree: [String: Any]) -> Void
  ) {
    let longEdgeCap = SupyDeviceTier.detect().ocrLongEdgeCap
    workQueue.async {
      let prepared = Self.downscaleIfNeeded(image, longEdgeCap: longEdgeCap)
      let tree = Self.recognizeStructured(
        in: prepared,
        languages: languages,
        includeElements: includeElements
      )
      DispatchQueue.main.async { completion(tree) }
    }
  }

  /// Warms the recognizer by running a request on a 1x1 image. Used by
  /// `prewarm()` in S3-06.
  func prewarm() {
    workQueue.async {
      let renderer = UIGraphicsImageRenderer(size: CGSize(width: 1, height: 1))
      let blank = renderer.image { ctx in
        UIColor.white.setFill()
        ctx.fill(CGRect(x: 0, y: 0, width: 1, height: 1))
      }
      _ = Self.recognizeText(in: blank, languages: ["en-US"])
    }
  }

  private static func recognizeText(
    in image: UIImage,
    languages: [String]
  ) -> String {
    guard let cgImage = image.cgImage else { return "" }
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    if !languages.isEmpty {
      request.recognitionLanguages = languages
    }
    let handler = VNImageRequestHandler(
      cgImage: cgImage,
      orientation: cgOrientation(from: image.imageOrientation),
      options: [:]
    )
    do {
      try handler.perform([request])
    } catch {
      return ""
    }
    guard let observations = request.results else { return "" }
    return observations
      .compactMap { $0.topCandidates(1).first?.string }
      .joined(separator: "\n")
  }

  /// Single OCR pass returning both the joined line text and every word's
  /// normalized box — the data the searchable-PDF exporter needs. Failures
  /// yield `("", [])` for the page rather than throwing.
  private static func recognizeTextAndWords(
    in image: UIImage,
    languages: [String]
  ) -> (text: String, words: [SupyOcrWord]) {
    guard let cgImage = image.cgImage else { return ("", []) }
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    if !languages.isEmpty {
      request.recognitionLanguages = languages
    }
    let handler = VNImageRequestHandler(
      cgImage: cgImage,
      orientation: cgOrientation(from: image.imageOrientation),
      options: [:]
    )
    do {
      try handler.perform([request])
    } catch {
      return ("", [])
    }
    guard let observations = request.results else { return ("", []) }
    var lineTexts: [String] = []
    var words: [SupyOcrWord] = []
    for obs in observations {
      guard let candidate = obs.topCandidates(1).first else { continue }
      lineTexts.append(candidate.string)
      words.append(contentsOf: wordBoxes(in: candidate, text: candidate.string))
    }
    return (lineTexts.joined(separator: "\n"), words)
  }

  private static func recognizeStructured(
    in image: UIImage,
    languages: [String],
    includeElements: Bool
  ) -> [String: Any] {
    let empty: [String: Any] = ["fullText": "", "blocks": []]
    guard let cgImage = image.cgImage else { return empty }
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    if !languages.isEmpty {
      request.recognitionLanguages = languages
    }
    let handler = VNImageRequestHandler(
      cgImage: cgImage,
      orientation: cgOrientation(from: image.imageOrientation),
      options: [:]
    )
    do {
      try handler.perform([request])
    } catch {
      return empty
    }
    guard let observations = request.results, !observations.isEmpty else {
      return empty
    }

    var lines: [[String: Any]] = []
    var lineTexts: [String] = []
    var union: CGRect?

    for obs in observations {
      guard let candidate = obs.topCandidates(1).first else { continue }
      let text = candidate.string
      lineTexts.append(text)
      union = union.map { $0.union(obs.boundingBox) } ?? obs.boundingBox

      var elements: [[String: Any]] = []
      if includeElements {
        elements = words(in: candidate, text: text)
      }
      lines.append([
        "text": text,
        "boundingBox": normRect(obs.boundingBox),
        "elements": elements,
      ])
    }

    let fullText = lineTexts.joined(separator: "\n")
    let block: [String: Any] = [
      "text": fullText,
      "boundingBox": normRect(union ?? .zero),
      "lines": lines,
    ]
    return ["fullText": fullText, "blocks": [block]]
  }

  /// Splits a recognized line into whitespace-delimited words and resolves
  /// each word's box via `VNRecognizedText.boundingBox(for:)`. Serialized shape
  /// for the structured OCR tree; wraps the shared [wordBoxes] tokenizer.
  private static func words(
    in candidate: VNRecognizedText,
    text: String
  ) -> [[String: Any]] {
    return wordBoxes(in: candidate, text: text).map { word in
      [
        "text": word.text,
        "boundingBox": [
          "left": word.left,
          "top": word.top,
          "width": word.width,
          "height": word.height,
        ],
      ]
    }
  }

  /// Shared tokenizer: splits [text] on whitespace and resolves each word's
  /// normalized `[0..1]` top-left box via `VNRecognizedText.boundingBox(for:)`.
  private static func wordBoxes(
    in candidate: VNRecognizedText,
    text: String
  ) -> [SupyOcrWord] {
    var out: [SupyOcrWord] = []
    var idx = text.startIndex
    while idx < text.endIndex {
      while idx < text.endIndex, text[idx].isWhitespace {
        idx = text.index(after: idx)
      }
      guard idx < text.endIndex else { break }
      let start = idx
      while idx < text.endIndex, !text[idx].isWhitespace {
        idx = text.index(after: idx)
      }
      let word = String(text[start..<idx])
      let box: CGRect
      if let rect = try? candidate.boundingBox(for: start..<idx)?.boundingBox {
        box = rect
      } else {
        box = .zero
      }
      let n = normRect(box)
      out.append(
        SupyOcrWord(
          text: word,
          left: n["left"] ?? 0,
          top: n["top"] ?? 0,
          width: n["width"] ?? 0,
          height: n["height"] ?? 0
        )
      )
    }
    return out
  }

  /// Normalizes a Vision bounding box (origin bottom-left) to Supy's
  /// top-left `[0..1]` `{left, top, width, height}` convention.
  private static func normRect(_ bb: CGRect) -> [String: Double] {
    return [
      "left": Double(bb.origin.x),
      "top": Double(1.0 - bb.origin.y - bb.size.height),
      "width": Double(bb.size.width),
      "height": Double(bb.size.height),
    ]
  }

  /// Downscales [image] so its long edge is at most [longEdgeCap]. The
  /// persisted page bytes returned to the consumer are unaffected — this
  /// only reduces the in-memory copy fed to Vision.
  private static func downscaleIfNeeded(
    _ image: UIImage,
    longEdgeCap: Int?
  ) -> UIImage {
    guard let cap = longEdgeCap else { return image }
    let longEdge = max(image.size.width, image.size.height) * image.scale
    guard longEdge > CGFloat(cap) else { return image }
    let scale = CGFloat(cap) / longEdge
    let targetSize = CGSize(
      width: image.size.width * scale,
      height: image.size.height * scale
    )
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = true
    let renderer = UIGraphicsImageRenderer(size: targetSize, format: format)
    return renderer.image { _ in
      image.draw(in: CGRect(origin: .zero, size: targetSize))
    }
  }

  private static func cgOrientation(
    from orientation: UIImage.Orientation
  ) -> CGImagePropertyOrientation {
    switch orientation {
    case .up: return .up
    case .down: return .down
    case .left: return .left
    case .right: return .right
    case .upMirrored: return .upMirrored
    case .downMirrored: return .downMirrored
    case .leftMirrored: return .leftMirrored
    case .rightMirrored: return .rightMirrored
    @unknown default: return .up
    }
  }
}

File: flutter/supy_scanner/android/src/main/kotlin/io/supy/scanner/document/OcrRunner.kt
package io.supy.scanner.document

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Rect
import android.net.Uri
import android.os.Handler
import android.os.Looper
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import io.supy.scanner.perf.DeviceTier
import java.io.IOException
import java.util.concurrent.Executors

/**
 * Runs ML Kit Text Recognition v2 (Latin script) over a list of page URIs and
 * concatenates the per-page text.
 *
 * Arabic OCR is intentionally not supported on Android — ML Kit's text
 * recognizer ships Latin, Chinese, Devanagari, Japanese, and Korean variants
 * only. Arabic-language scans on Android will return an empty `ocrText` for
 * the Arabic portions; the iOS side covers `ar` via `VNRecognizeTextRequest`.
 * See `docs/MIGRATION.md` for the Scanbot-vs-supy_scanner coverage table.
 */
internal class OcrRunner {

    private val recognizer: TextRecognizer =
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    /** Off-main-thread decode of the source image before handing it to ML Kit. */
    private val ioExecutor = Executors.newSingleThreadExecutor()

    /**
     * Reads each [Uri], decodes its pixel dimensions, runs OCR, and invokes
     * [onComplete] on the calling thread with an enriched page list and the
     * concatenated text. Always invokes [onComplete] exactly once.
     */
    fun run(
        context: Context,
        pages: List<PageReencoder.ReencodedPage>,
        onComplete: (pages: List<Map<String, Any?>>, ocrText: String) -> Unit,
    ) {
        if (pages.isEmpty()) {
            onComplete(emptyList(), "")
            return
        }

        val results = arrayOfNulls<PageOcr>(pages.size)
        var remaining = pages.size

        val longEdgeCap = DeviceTier.detect(context).ocrLongEdgeCap()

        pages.forEachIndexed { index, page ->
            val uri = page.uri
            val (width, height) = readDimensions(context, uri)
            val image = try {
                buildInputImage(context, uri, width, height, longEdgeCap)
            } catch (e: IOException) {
                results[index] = PageOcr(uri, width, height, "", page.quality, page.qualityScore)
                remaining -= 1
                if (remaining == 0) emit(results, onComplete)
                return@forEachIndexed
            }

            recognizer.process(image)
                .addOnSuccessListener { recognized ->
                    results[index] = PageOcr(uri, width, height, recognized.text, page.quality, page.qualityScore)
                }
                .addOnFailureListener {
                    // Per-page failure → empty text for that page, partial
                    // result for the rest. Don't surface an error code.
                    results[index] = PageOcr(uri, width, height, "", page.quality, page.qualityScore)
                }
                .addOnCompleteListener {
                    remaining -= 1
                    if (remaining == 0) emit(results, onComplete)
                }
        }
    }

    /**
     * Like [run], but also collects per-page word boxes for the searchable-PDF
     * text layer (v1.2 Phase DC8). One ML Kit pass per page — the same pass
     * that produces `ocrText` — so `outputFormat=searchablePdf` costs no extra
     * OCR over a plain PDF run. [onComplete] fires exactly once, on the calling
     * thread, with a `wordsPerPage` list index-aligned to the emitted pages.
     */
    fun runWithWords(
        context: Context,
        pages: List<PageReencoder.ReencodedPage>,
        onComplete: (
            pages: List<Map<String, Any?>>,
            ocrText: String,
            wordsPerPage: List<List<PdfAssembler.Word>>,
        ) -> Unit,
    ) {
        if (pages.isEmpty()) {
            onComplete(emptyList(), "", emptyList())
            return
        }

        val results = arrayOfNulls<PageOcrWords>(pages.size)
        var remaining = pages.size
        val longEdgeCap = DeviceTier.detect(context).ocrLongEdgeCap()

        pages.forEachIndexed { index, page ->
            val uri = page.uri
            val (width, height) = readDimensions(context, uri)
            val image = try {
                buildInputImage(context, uri, width, height, longEdgeCap)
            } catch (e: IOException) {
                results[index] = PageOcrWords(uri, width, height, "", page.quality, page.qualityScore, emptyList())
                remaining -= 1
                if (remaining == 0) emitWords(results, onComplete)
                return@forEachIndexed
            }
            val imgWidth = image.width
            val imgHeight = image.height

            recognizer.process(image)
                .addOnSuccessListener { recognized ->
                    val words = collectWords(recognized, imgWidth, imgHeight)
                    results[index] =
                        PageOcrWords(uri, width, height, recognized.text, page.quality, page.qualityScore, words)
                }
                .addOnFailureListener {
                    results[index] =
                        PageOcrWords(uri, width, height, "", page.quality, page.qualityScore, emptyList())
                }
                .addOnCompleteListener {
                    remaining -= 1
                    if (remaining == 0) emitWords(results, onComplete)
                }
        }
    }

    /** Flattens ML Kit elements ("words") into normalized [PdfAssembler.Word]s. */
    private fun collectWords(
        text: com.google.mlkit.vision.text.Text,
        imgWidth: Int,
        imgHeight: Int,
    ): List<PdfAssembler.Word> {
        val words = mutableListOf<PdfAssembler.Word>()
        for (block in text.textBlocks) {
            for (line in block.lines) {
                for (element in line.elements) {
                    val box = normRect(element.boundingBox, imgWidth, imgHeight)
                    words.add(
                        PdfAssembler.Word(
                            text = element.text,
                            left = box.getValue("left"),
                            top = box.getValue("top"),
                            width = box.getValue("width"),
                            height = box.getValue("height"),
                        ),
                    )
                }
            }
        }
        return words
    }

    private fun emitWords(
        results: Array<PageOcrWords?>,
        onComplete: (List<Map<String, Any?>>, String, List<List<PdfAssembler.Word>>) -> Unit,
    ) {
        val present = results.filterNotNull()
        val pages = present.map { page ->
            buildMap<String, Any?> {
                put("uri", page.uri.toString())
                put("width", page.width)
                put("height", page.height)
                if (page.quality != null) put("quality", page.quality)
                if (page.qualityScore != null) put("qualityScore", page.qualityScore)
            }
        }
        val text = present.joinToString(separator = "\n\n") { it.text }.trim()
        val wordsPerPage = present.map { it.words }
        onComplete(pages, text, wordsPerPage)
    }

    /**
     * Runs a single-image structured OCR pass and posts the block → line →
     * element tree consumed by `SupyRecognizedText.fromMap` on the Dart side.
     *
     * ML Kit exposes `textBlocks → lines → elements`, each with a pixel-space
     * `Rect`. Boxes are normalized to `[0..1]` (origin top-left) against the
     * source image dimensions to match `SupyBarcode.boundingBox`. Both
     * callbacks are always invoked exactly once, on the main thread.
     *
     * The `languages` arg is accepted for wire symmetry with iOS but has no
     * effect here — ML Kit's bundled recognizer is Latin-only (see the class
     * doc and `docs/MIGRATION.md`).
     */
    fun recognizeStructured(
        context: Context,
        uri: Uri,
        includeElements: Boolean,
        onComplete: (Map<String, Any?>) -> Unit,
        onError: (code: String, message: String) -> Unit,
    ) {
        val main = Handler(Looper.getMainLooper())
        ioExecutor.execute {
            val (width, height) = readDimensions(context, uri)
            val longEdgeCap = DeviceTier.detect(context).ocrLongEdgeCap()
            val image = try {
                buildInputImage(context, uri, width, height, longEdgeCap)
            } catch (e: IOException) {
                main.post { onError("model_unavailable", "recognizeText: could not decode $uri (${e.message})") }
                return@execute
            }
            // ML Kit sizes its geometry against the (possibly downscaled)
            // InputImage, so normalize against that — not the on-disk dims.
            val imgWidth = image.width
            val imgHeight = image.height
            recognizer.process(image)
                .addOnSuccessListener { recognized ->
                    val tree = buildTree(recognized, imgWidth, imgHeight, includeElements)
                    onComplete(tree)
                }
                .addOnFailureListener { e ->
                    onError("model_unavailable", "recognizeText: ML Kit failed (${e.message})")
                }
        }
    }

    private fun buildTree(
        text: com.google.mlkit.vision.text.Text,
        imgWidth: Int,
        imgHeight: Int,
        includeElements: Boolean,
    ): Map<String, Any?> {
        val blocks = text.textBlocks.map { block ->
            val lines = block.lines.map { line ->
                val elements = if (includeElements) {
                    line.elements.map { element ->
                        mapOf(
                            "text" to element.text,
                            "boundingBox" to normRect(element.boundingBox, imgWidth, imgHeight),
                        )
                    }
                } else {
                    emptyList()
                }
                mapOf(
                    "text" to line.text,
                    "boundingBox" to normRect(line.boundingBox, imgWidth, imgHeight),
                    "elements" to elements,
                )
            }
            mapOf(
                "text" to block.text,
                "boundingBox" to normRect(block.boundingBox, imgWidth, imgHeight),
                "lines" to lines,
            )
        }
        return mapOf("fullText" to text.text, "blocks" to blocks)
    }

    /**
     * Normalizes a pixel-space [Rect] to Supy's `[0..1]` top-left
     * `{left, top, width, height}` convention. A null box (ML Kit may omit it)
     * or a degenerate image size collapses to the zero rect.
     */
    private fun normRect(rect: Rect?, imgWidth: Int, imgHeight: Int): Map<String, Double> {
        if (rect == null || imgWidth <= 0 || imgHeight <= 0) {
            return mapOf("left" to 0.0, "top" to 0.0, "width" to 0.0, "height" to 0.0)
        }
        val w = imgWidth.toDouble()
        val h = imgHeight.toDouble()
        return mapOf(
            "left" to rect.left / w,
            "top" to rect.top / h,
            "width" to rect.width() / w,
            "height" to rect.height() / h,
        )
    }

    fun close() {
        recognizer.close()
        ioExecutor.shutdown()
    }

    private fun emit(
        results: Array<PageOcr?>,
        onComplete: (List<Map<String, Any?>>, String) -> Unit,
    ) {
        val pages = results.filterNotNull().map { page ->
            buildMap<String, Any?> {
                put("uri", page.uri.toString())
                put("width", page.width)
                put("height", page.height)
                if (page.quality != null) put("quality", page.quality)
                if (page.qualityScore != null) put("qualityScore", page.qualityScore)
            }
        }
        val text = results.filterNotNull()
            .joinToString(separator = "\n\n") { it.text }
            .trim()
        onComplete(pages, text)
    }

    /**
     * Returns an `InputImage` for [uri], downscaled to roughly [longEdgeCap]
     * pixels on the long edge when a cap applies and the page is larger than
     * it. The persisted JPEG on disk is left untouched — only the in-memory
     * copy fed to ML Kit is reduced.
     */
    private fun buildInputImage(
        context: Context,
        uri: Uri,
        width: Int,
        height: Int,
        longEdgeCap: Int?,
    ): InputImage {
        val longEdge = maxOf(width, height)
        if (longEdgeCap == null || longEdge <= 0 || longEdge <= longEdgeCap) {
            return InputImage.fromFilePath(context, uri)
        }

        var sampleSize = 1
        while (longEdge / (sampleSize * 2) >= longEdgeCap) {
            sampleSize *= 2
        }
        val options = BitmapFactory.Options().apply {
            inSampleSize = sampleSize
            inPreferredConfig = Bitmap.Config.ARGB_8888
        }
        val bitmap = context.contentResolver.openInputStream(uri)?.use { stream ->
            BitmapFactory.decodeStream(stream, null, options)
        } ?: throw IOException("Failed to decode $uri for OCR downscale")
        return InputImage.fromBitmap(bitmap, 0)
    }

    private fun readDimensions(context: Context, uri: Uri): Pair<Int, Int> {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        try {
            context.contentResolver.openInputStream(uri)?.use { stream ->
                BitmapFactory.decodeStream(stream, null, options)
            }
        } catch (_: IOException) {
            return 0 to 0
        }
        return options.outWidth to options.outHeight
    }

    private data class PageOcr(
        val uri: Uri,
        val width: Int,
        val height: Int,
        val text: String,
        val quality: String?,
        val qualityScore: Double?,
    )

    private data class PageOcrWords(
        val uri: Uri,
        val width: Int,
        val height: Int,
        val text: String,
        val quality: String?,
        val qualityScore: Double?,
        val words: List<PdfAssembler.Word>,
    )
}
